### ISSUE TITLE
Vscode crash on kernel restart or interrupt

### ISSUE THREAD

**Issue by IsaacWarren** [microsoft/vscode-jupyter#16283]
### Applies To

- [x] Notebooks (.ipynb files)
- [ ] Interactive Window and\/or Cell Scripts (.py files with \#%% markers)

### What happened?

When a kernel that has used mpi4py.MPI.Spawn is restarted VSCode crashes. I looked through the logs myself and nothing obvious was found. This only happens if the kernel is launched from VSCode, not if connecting to an external jupyter server. This is not reproducible in either jupyter lab or Cursor's notebook support, both restart the kernel successfully as expected. Reproduced in a fresh install of VSCode insiders with only the jupyter extensions and its dependencies installed.
### Reproducer
```python
import mpi4py.MPI as MPI
comm_world = MPI.COMM_WORLD
comm_world.Spawn(
            # get the same python executable that is currently running
            "python", ["-c",  "import mpi4py.MPI as MPI; print(MPI.COMM_WORLD.Get_rank())"],
            1,
        )
```

### Installed packages
Packages installed:
```
pip install mpi4py jupyter
pip install -i https://pypi.anaconda.org/mpi4py/simple mpich 
```

Python version: Python 3.12.7
Output of pip list:
```
Package                    Version                  Editable project location
-------------------------- ------------------------ ----------------------------
Package                   Version
------------------------- --------------
anyio                     4.7.0
appnope                   0.1.4
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 3.0.0
async-lru                 2.0.4
attrs                     24.2.0
babel                     2.16.0
beautifulsoup4            4.12.3
bleach                    6.2.0
certifi                   2024.8.30
cffi                      1.17.1
charset-normalizer        3.4.0
comm                      0.2.2
debugpy                   1.8.9
decorator                 5.1.1
defusedxml                0.7.1
executing                 2.1.0
fastjsonschema            2.21.1
fqdn                      1.5.1
h11                       0.14.0
httpcore                  1.0.7
httpx                     0.28.1
idna                      3.10
ipykernel                 6.29.5
ipython                   8.30.0
ipywidgets                8.1.5
isoduration               20.11.0
jedi                      0.19.2
Jinja2                    3.1.4
json5                     0.10.0
jsonpointer               3.0.0
jsonschema                4.23.0
jsonschema-specifications 2024.10.1
jupyter                   1.1.1
jupyter_client            8.6.3
jupyter-console           6.6.3
jupyter_core              5.7.2
jupyter-events            0.10.0
jupyter-lsp               2.2.5
jupyter_server            2.14.2
jupyter_server_terminals  0.5.3
jupyterlab                4.3.3
jupyterlab_pygments       0.3.0
jupyterlab_server         2.27.3
jupyterlab_widgets        3.0.13
MarkupSafe                3.0.2
matplotlib-inline         0.1.7
mistune                   3.0.2
mpi4py                    4.0.1
mpich                     4.2.3
nbclient                  0.10.1
nbconvert                 7.16.4
nbformat                  5.10.4
nest-asyncio              1.6.0
notebook                  7.3.1
notebook_shim             0.2.4
overrides                 7.7.0
packaging                 24.2
pandocfilters             1.5.1
parso                     0.8.4
pexpect                   4.9.0
pip                       24.2
platformdirs              4.3.6
prometheus_client         0.21.1
prompt_toolkit            3.0.48
psutil                    6.1.0
ptyprocess                0.7.0
pure_eval                 0.2.3
pycparser                 2.22
Pygments                  2.18.0
python-dateutil           2.9.0.post0
python-json-logger        2.0.7
PyYAML                    6.0.2
pyzmq                     26.2.0
referencing               0.35.1
requests                  2.32.3
rfc3339-validator         0.1.4
rfc3986-validator         0.1.1
rpds-py                   0.22.3
Send2Trash                1.8.3
setuptools                75.6.0
six                       1.17.0
sniffio                   1.3.1
soupsieve                 2.6
stack-data                0.6.3
terminado                 0.18.1
tinycss2                  1.4.0
tornado                   6.4.2
traitlets                 5.14.3
types-python-dateutil     2.9.0.20241206
typing_extensions         4.12.2
uri-template              1.3.0
urllib3                   2.2.3
wcwidth                   0.2.13
webcolors                 24.11.1
webencodings              0.5.1
websocket-client          1.8.0
widgetsnbextension        4.0.13

```

### VS Code Version

Version: 1.95.3 (Universal) Commit: f1a4fb101478ce6ec82fe9627c43efbf9e98c813 Date: 2024-11-13T14:50:04.152Z Electron: 32.2.1 ElectronBuildId: 10427718 Chromium: 128.0.6613.186 Node.js: 20.18.0 V8: 12.8.374.38-electron.0 OS: Darwin arm64 22.6.0

### Jupyter Extension Version

v2024.10.0

### Jupyter logs

```shell
Visual Studio Code (1.95.3, undefined, desktop)
Jupyter Extension Version: 2024.10.0.
Python Extension Version: 2024.20.0.
Pylance Extension Version: 2024.12.1.
Platform: darwin (arm64).
Temp Storage folder ~/Library/Application Support/Code/User/globalStorage/ms-toolsai.jupyter/version-2024.10.0
Workspace folder ~/Bodo/Bodo, Home = /Users/isaacwarren
15:46:13.819 [debug] Start refreshing Kernel Picker (1733867173819)
15:46:13.867 [debug] Get Custom Env Variables, Class name = Em, completed in 47ms, has a truthy return value, Arg 1: undefined, Arg 2: "RunPythonCode"
15:46:13.867 [debug] Jupyter Paths /kernels: 
15:46:13.867 [debug] Kernel Spec Root Paths, /usr/share/jupyter/kernels, /usr/local/share/jupyter/kernels, ~/Library/Jupyter/kernels
15:46:13.876 [debug] Start refreshing Interpreter Kernel Picker
15:46:14.242 [debug] KernelProvider switched kernel to id = .jvsc74a57bd0b1707b3cea6f2a9def084b805f20f1073070644042747105ee0afdbbd3d37920./Users/~/Bodo/Bodo/venv/python./Users/~/Bodo/Bodo/venv/python.-m#ipykernel_launcher
15:46:14.243 [debug] start the kernel, options.disableUI=true for ~/Bodo/Bodo/test.ipynb
15:46:14.260 [debug] Controller selection change completed
15:46:14.287 [debug] Get Custom Env Variables, Class name = Em, completed in 0ms, has a truthy return value, Arg 1: "~/Bodo/Bodo", Arg 2: "RunPythonCode"
15:46:14.288 [info] Starting Kernel (Python Path: ~/Bodo/Bodo/venv/bin/python, Venv, 3.12.7) for '~/Bodo/Bodo/test.ipynb' (disableUI=true)
15:46:14.306 [debug] Get Custom Env Variables, Class name = Em, completed in 0ms, has a truthy return value, Arg 1: "~/Bodo/Bodo/test.ipynb", Arg 2: "RunPythonCode"
15:46:14.368 [debug] Launching kernel .jvsc74a57bd0b1707b3cea6f2a9def084b805f20f1073070644042747105ee0afdbbd3d37920./Users/~/Bodo/Bodo/venv/python./Users/~/Bodo/Bodo/venv/python.-m#ipykernel_launcher for ~/Bodo/Bodo/test.ipynb in //Users/~/Bodo/Bodo with ports 9001, 9000, 9004, 9002, 9003
15:46:14.481 [debug] Got env vars from Python Ext in 175ms for ~/Bodo/Bodo/venv/bin/python, with env var count 56.
15:46:14.481 [debug] Getting activated env variables, Class name = Id, completed in 175ms, has a truthy return value, Arg 1: "~/Bodo/Bodo/test.ipynb", Arg 2: "~/Bodo/Bodo/venv/bin/python", Arg 3: undefined
15:46:14.483 [info] Process Execution: ~/Bodo/Bodo/venv/bin/python -c "import ipykernel; print(ipykernel.__version__); print("5dc3a68c-e34e-4080-9c3e-2a532b2ccb4d"); print(ipykernel.__file__)"
15:46:14.483 [info] Process Execution: ~/Bodo/Bodo/venv/bin/python -m ipykernel_launcher --f=/Users/~/Library/Jupyter/runtime/kernel-v3f20346ef0898e34428f6206437e2eefef60bbe70.json
    > cwd: //Users/~/Bodo/Bodo
15:46:14.483 [debug] Kernel process 49848.
15:46:14.494 [debug] Got env vars from Python Ext in 207ms for ~/Bodo/Bodo/venv/bin/python, with env var count 56.
15:46:14.494 [debug] Getting activated env variables, Class name = Id, completed in 207ms, has a truthy return value, Arg 1: "~/Bodo/Bodo", Arg 2: "~/Bodo/Bodo/venv/bin/python", Arg 3: undefined
15:46:14.495 [info] Process Execution: ~/Bodo/Bodo/venv/bin/python -m pip list
15:46:15.649 [debug] Kernel output 49848: To connect another client to this kernel, use:
--existing kernel-v3f20346ef0898e34428f6206437e2eefef60bbe70.json
15:46:15.669 [debug] Waiting for Raw Session to be ready in postStartRawSession
15:46:15.669 [debug] Waiting for Raw session to be ready, status: connected
15:46:15.669 [debug] Waited for Raw session to be ready & got status: connected
15:46:15.669 [debug] Successfully waited for Raw Session to be ready in postStartRawSession
15:46:15.669 [debug] Kernel status is 'unknown' before requesting kernel info and after ready
15:46:15.669 [debug] Sending request for kernelInfo
15:46:15.715 [debug] Successfully completed postStartRawSession after 1 attempt(s) in 46ms
15:46:15.860 [debug] Interpreter for Pylance for Notebook URI "~/Bodo/Bodo/test.ipynb" is ~/Bodo/Bodo/venv/bin/python
15:46:15.893 [debug] Requesting Kernel info
15:46:15.894 [info] Kernel successfully started
15:46:15.895 [debug] getDataDirsImpl, Class name = Kr (started execution), Arg 1: "~/Bodo/Bodo/test.ipynb", Arg 2: "~/Bodo/Bodo/venv/bin/python"
15:46:15.895 [debug] Get Custom Env Variables, Class name = Em, completed in 0ms, has a truthy return value, Arg 1: undefined, Arg 2: "RunPythonCode"
15:46:15.896 [debug] Jupyter Paths : 
15:46:15.897 [info] Process Execution: ~/Bodo/Bodo/venv/bin/python /Users/~/.vscode/extensions/ms-toolsai.jupyter-2024.10.0-darwin-arm64/pythonFiles/printJupyterDataDir.py
15:46:15.920 [debug] Got an empty Jupyter Data Dir from /Users/~/Bodo/Bodo/venv/bin/python, stderr = undefined
15:46:15.920 [debug] getDataDirsImpl, Class name = Kr, completed in 25ms, has a truthy return value, Arg 1: "~/Bodo/Bodo/test.ipynb", Arg 2: "~/Bodo/Bodo/venv/bin/python", Return Value: <Uri:/Users/~/Library/Jupyter>, <Uri:/Users/~/Bodo/Bodo/venv/share/jupyter>, <Uri:/usr/local/share/jupyter>, <Uri:/usr/share/jupyter>
15:46:16.706 [debug] Controller startUsingPythonInterpreter:'.jvsc74a57bd05edc29c2ed010d6458d71a83433b383a96a8cbd3efe8531bc90c4b8a5b8bcec9./Library/Developer/CommandLineTools/usr/bin/python3./Library/Developer/CommandLineTools/usr/bin/python3.-m#ipykernel_launcher' for view = 'jupyter-notebook' is no longer a valid
15:46:16.706 [debug] Controller startUsingPythonInterpreter:'.jvsc74a57bd05edc29c2ed010d6458d71a83433b383a96a8cbd3efe8531bc90c4b8a5b8bcec9./Library/Developer/CommandLineTools/usr/bin/python3./Library/Developer/CommandLineTools/usr/bin/python3.-m#ipykernel_launcher (Interactive)' for view = 'interactive' is no longer a valid
15:46:16.706 [warn] Disposing old controller startUsingPythonInterpreter:'.jvsc74a57bd05edc29c2ed010d6458d71a83433b383a96a8cbd3efe8531bc90c4b8a5b8bcec9./Library/Developer/CommandLineTools/usr/bin/python3./Library/Developer/CommandLineTools/usr/bin/python3.-m#ipykernel_launcher' for view = 'jupyter-notebook'
15:46:16.706 [debug] Disposing controller .jvsc74a57bd05edc29c2ed010d6458d71a83433b383a96a8cbd3efe8531bc90c4b8a5b8bcec9./Library/Developer/CommandLineTools/usr/bin/python3./Library/Developer/CommandLineTools/usr/bin/python3.-m#ipykernel_launcher associated with connection .jvsc74a57bd05edc29c2ed010d6458d71a83433b383a96a8cbd3efe8531bc90c4b8a5b8bcec9./Library/Developer/CommandLineTools/usr/bin/python3./Library/Developer/CommandLineTools/usr/bin/python3.-m#ipykernel_launcher 
15:46:16.714 [warn] Disposing old controller startUsingPythonInterpreter:'.jvsc74a57bd05edc29c2ed010d6458d71a83433b383a96a8cbd3efe8531bc90c4b8a5b8bcec9./Library/Developer/CommandLineTools/usr/bin/python3./Library/Developer/CommandLineTools/usr/bin/python3.-m#ipykernel_launcher (Interactive)' for view = 'interactive'
15:46:16.714 [debug] Disposing controller .jvsc74a57bd05edc29c2ed010d6458d71a83433b383a96a8cbd3efe8531bc90c4b8a5b8bcec9./Library/Developer/CommandLineTools/usr/bin/python3./Library/Developer/CommandLineTools/usr/bin/python3.-m#ipykernel_launcher (Interactive) associated with connection .jvsc74a57bd05edc29c2ed010d6458d71a83433b383a96a8cbd3efe8531bc90c4b8a5b8bcec9./Library/Developer/CommandLineTools/usr/bin/python3./Library/Developer/CommandLineTools/usr/bin/python3.-m#ipykernel_launcher 
15:46:16.723 [warn] No interpreter with path /opt/homebrew/anaconda3/bin/python found in Python API, will convert Uri path to string as Id /opt/homebrew/anaconda3/bin/python
15:46:16.724 [debug] Refreshed Environments
15:46:18.918 [warn] Kernel Spec for 'Python 3 (ipykernel)' (~/Library/Jupyter/kernels/bodo_platform_dummy_kernel/kernel.json) hidden, as we cannot find a matching interpreter argv = '/opt/homebrew/anaconda3/bin/python'. To resolve this, please change '/opt/homebrew/anaconda3/bin/python' to point to the fully qualified Python executable.
15:46:18.918 [debug] End refreshing Kernel Picker (1733867173819)
15:51:29.769 [debug] Handle Execution of Cells 0 for ~/Bodo/Bodo/test.ipynb
15:51:29.775 [debug] Kernel acknowledged execution of cell 0 @ 1733867489775
15:51:29.837 [debug] Kernel output 49848: 0

15:51:31.565 [debug] Restart kernel command handler for ~/Bodo/Bodo/test.ipynb
15:51:31.566 [debug] restart the kernel, options.disableUI=false for ~/Bodo/Bodo/test.ipynb
15:51:31.566 [info] Restart requested ~/Bodo/Bodo/test.ipynb
15:51:31.566 [debug] Cancel pending cells
15:51:31.567 [debug] Dispose Kernel process 49848.
15:51:31.570 [debug] Get Custom Env Variables, Class name = Em, completed in 0ms, has a truthy return value, Arg 1: "~/Bodo/Bodo/test.ipynb", Arg 2: "RunPythonCode"
15:51:31.570 [debug] Got env vars from Python Ext in 0ms for ~/Bodo/Bodo/venv/bin/python, with env var count 56.
15:51:31.570 [debug] Getting activated env variables, Class name = Id, completed in 0ms, has a truthy return value, Arg 1: "~/Bodo/Bodo/test.ipynb", Arg 2: "~/Bodo/Bodo/venv/bin/python", Arg 3: undefined
15:51:31.571 [info] Process Execution: ~/Bodo/Bodo/venv/bin/python -c "import ipykernel; print(ipykernel.__version__); print("5dc3a68c-e34e-4080-9c3e-2a532b2ccb4d"); print(ipykernel.__file__)"
15:51:31.575 [debug] Launching kernel .jvsc74a57bd0b1707b3cea6f2a9def084b805f20f1073070644042747105ee0afdbbd3d37920./Users/~/Bodo/Bodo/venv/python./Users/~/Bodo/Bodo/venv/python.-m#ipykernel_launcher for ~/Bodo/Bodo/test.ipynb in //Users/~/Bodo/Bodo with ports 9006, 9005, 9009, 9007, 9008
15:51:31.578 [info] Process Execution: ~/Bodo/Bodo/venv/bin/python -m ipykernel_launcher --f=/Users/~/Library/Jupyter/runtime/kernel-v38d207f7b09f2b2ff5262dd736fbd4ae1f1f35bee.json
    > cwd: //Users/~/Bodo/Bodo
15:51:31.578 [debug] Kernel process 51885.
```

### Coding Language and Runtime Version

Python 3.12.7

### Language Extension Version (if applicable)

v2024.20.0

### Anaconda Version (if applicable)

NA

### Running Jupyter locally or remotely?

Local

**Comment by DonJayamanne** [comment]
Thank you for filing this issue, yes I can replicate this at my end too.

**Comment by DonJayamanne** [comment]
@IsaacWarren I found that with Jupyter lab the kernel process crashes when you attempt to interrupt the kernel.
However Jupyter lab application doesn't crash itself. Its because the kernel process is spawned as a detached process, hence it doesn't crash the parent process.
I'm not sure about cursor, however given the fact taht they too are built on VS Code, I'm surprised it doesn't crash, that they'd be using the same code/extensions. Unless they have disabled a feature in jupyter extension that always starts Jupyter servers (i.e. starting jupyter lab server in order to start the kernels), else Cursor too would crash.

In fact this isn't limited to just VS Code, zed too crashes with this same code.
The only work around I have at this stage is to add the following setting into your settings
```json
"jupyter.disableZMQSupport": true
```
You will see some red squiggles indicating this isn't a valid setting, ignore that, reload VS Code and try again.
This time you will be promtped to select an interpreter an probably install Jupyter application, 
Once thats done, you should be unblocked.

Please note, this is merely a work around for now.

**Comment by ehsantn** [comment]
@DonJayamanne thanks for your response. After a bit of testing, sounds to me like Python's `atexit` functions are not executed properly during restart. Cleaning up MPI spawned processes manually prevents the crash. Can you think of any issues around the kernel tear down and `atexit` functions?

**Comment by DonJayamanne** [comment]
> , sounds to me like Python's atexit functions are not executed properly during restart. 

Not sure why you think thats the case, from what I know about IPykernel and the related packages they are doing the right thing, those packages have been around for a while and battle tested. given the fact that this package is causing issues in 3 applications, I suspect its caused by this the code in this package.
I.e. sending a SIGINT should not cause the process to crash, however thats the case.
I.e. when I enumerate the child processes spawned by the python code and send a SIGINT to one of those process, the kernel process also crashes, i.e. the child is causing the parent to crash, ideally it should not crash at all, SIGTERM or similar signals can cause an exit, not SIGINT.

Found more details on the root cause.
See here https://github.com/mpi4py/mpi4py/issues/522

Given the fact that this is caused by the package itself, and SIGINT signal isn't being adhered to correctly, I'm closing this as out of scope.

**Comment by IsaacWarren** [comment]
@DonJayamanne Thanks for taking a look. The reason we suspected some atexits might not be called is because mpi4py calls MPI.Finalize() in an at exit (ref: https://github.com/mpi4py/mpi4py/blob/50313c2098bc345eb4fac78617185429dacc3e96/docs/source/overview.rst?plain=1#L414) and adding MPI.Finalize() at the end of the cell prevents the crash.
As to the mpi4py issue linked we don't use openMPI in the reproducer, we use mpich which doesn't have the same issue according to the thread. Something we found when looking through the code is https://github.com/microsoft/vscode-jupyter/blob/f80bf701a710328b20c5931d621e8d83813055ea/src/kernels/raw/launcher/kernelProcess.node.ts#L355 which seems to send a SIGTERM to the kernel if it hasn't shutdown cleanly within 1 second? Our hypothesis was the SIGTERM is being sent before all atexit handlers can run leading to a shutdown without all cleanup. We wanted to test it ourselves by playing with the cutoff but are struggling to get the extension to build for us locally (`npm ci` seems to spin indefinitely  while utf-8-validate@5.0.9, zeromq@6.0.0-beta.6, and bufferutil@4.0.6). If you have anymore insight into this issue it would be greatly appreciated.

**Comment by DonJayamanne** [comment]
Please note, as mentioned even with Jupyter Lab application, if i send a SIGINT to one of the child processes, the kernel process crashes

> Our hypothesis was the SIGTERM is being sent before all atexit handlers can run leading to a shutdown without all cleanup.

I.e. there's no SIGTERM being sent at all in this case. 
Just start Jupyter lab, run a cell with the mpi code that spawns the sub processes.
next get the child processes and send a SIGINT from a terminal to one fo the child procs and you'll run into the issue i've mentioned above.
